client/app/components/items-list/classes/ItemsSource.js

```diff
@@ -0,0 +1,247 @@
+  };
+
+  toggleSorting = (orderByField) => {
+    this._sorter.toggleField(orderByField);
+    this._savedOrderByField = this._sorter.field;
+    this._changed({ sorting: true });
+  };
+
+  updateSearch = (searchTerm) => {
+    // here we update state directly, but later `fetchData` will update it properly
+    this._searchTerm = searchTerm;
+    // in search mode ignore the ordering and use the ranking order
+    // provided by the server-side FTS backend instead, unless it was
+    // requested by the user by actively ordering in search mode
+    if (searchTerm === '') {
+      this._sorter.setField(this._savedOrderByField); // restore ordering
+    } else {
+      this._sorter.setField(null);
+    }
+    this._paginator.setPage(1);
+    this._changed({ search: true, pagination: { page: true } });
+  };
+
+  updateSelectedTags = (selectedTags) => {
+    this._selectedTags = selectedTags;
+    this._paginator.setPage(1);
+    this._changed({ tags: true, pagination: { page: true } });
+  };
+
+  update = () => this._doRequest(this._getRequest());
+
+  handleError(error) {
+    if (isFunction(this.onError)) {
+      this.onError(error);
+    }
+  }
+}
+
+export class PlainItemsSource extends ItemsSource {
+  _items = [];
+
+  // eslint-disable-next-line class-methods-use-this
+  _processResults(data) {
+    return {
+      results: data,
+      count: data.length,
+    };
+  }
+
+  _updateResults(items, totalCount) {
+    this._items = this._sorter.sort(items);
+    this._pageItems = this._paginator.getItemsForPage(this._items);
+    this._paginator.setTotalCount(totalCount);
+  }
+
+  _changed({ sorting, pagination: { page } }) {
+    if (sorting || page) {
+      this._beforeUpdate().then(() => {
+        if (sorting) {
+          this._items = this._sorter.sort(this._items);
+        }
+        this._pageItems = this._paginator.getItemsForPage(this._items);
+        this._afterUpdate();
+      });
+    } else {
+      this.update();
+    }
+  }
+}
+
+export class ResourceItemsSource extends PlainItemsSource {
+  static _processPlainList(results, processItem, context) {
+    return map(results, item => processItem(item, context));
+  }
+
+  static _processPaginatedList({ results, count }, processItem, context) {
+    return {
+      results: map(results, item => processItem(item, context)),
+      count,
+    };
+  }
+
+  static _createResourceFetcher(getResource, getItemProcessor, processResults) {
+    getItemProcessor = isFunction(getItemProcessor) ? getItemProcessor : () => identity;
+
+    return (request, context) => {
+      const resource = getResource(context);
+
+      let processItem = getItemProcessor(context);
+      processItem = isFunction(processItem) ? processItem : item => item;
+
+      return resource(request).$promise.then(data => processResults(data, processItem, context));
+    };
+  }
+
+  _processResults(data) {
+    if (this._isPlainList) {
+      return super._processResults(data); // from PlainItemsSource
+    }
```

Comment:
Hmm.. something about this doesn't sit well with me.
If the `ResourceItemsSource` extends `PlainItemsSource`, why would it need to be told if it should use inherited methods or not.

Could this be avoided by the implementation using the appropriate `ItemsSource`?
```js
itemsList(
  getRequest(request, { currentPage }) {
    new PlainItemsSource({
      ...
```

Reply:
Good point. I'll re-think this :+1:

Reply:
@ranbena I was thinking about it for a long time and ended up with merging `ItemsSource` and `PlainItemsSource` classes - there were really minor differences between them (technically they're different enough, but I decided just to add flag). Also it allowed to simplify `ResourceItemsSource` a lot.

Reply:
@ranbena Finally decided to implement pattern "strategy" for fetching items and I'm pretty happy with the results :slightly_smiling_face:

Reply:
On it